keps/sig-network/1880-multiple-service-cidrs/README.md

```diff
@@ -0,0 +1,954 @@
+  - ServiceCIDR immutability
+  - default IPFamily
+  - two or one IP families per ServiceCIDRConfig
+- Gather feedback from developers and users
+- Document and add more complex testing scenarios: scaling out ServiceCIDRConfigs, ...
+- Additional tests are in Testgrid and linked in KEP
+- Scale test to O(10K) services and O(1K) ranges
+- Improve performance on the allocation logic, O(1) for allocating a free address
+- Allowing time for feedback
+
+#### GA
+
+- 2 examples of real-world usage
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and GA/stable, because there's
+no opportunity for user feedback, or even bug reports, in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include [conformance
+tests].**
+
+[conformance tests]:
+    https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+### Upgrade / Downgrade / Version Skew Strategy
+
+The source of truth are the IPs assigned to the Services, both the old and new methods have
+reconcile loops that rebuild the state of the allocators based on the assigned IPs to the Services,
+this allows to support upgrades and skewed clusters.
+
+Clusters running with the new allocation model will not keep running the reconcile loops that
+keep the bitmap used by the allocators.
+
+Since the new allocation model will remove some of the limitations of the current model, skewed
+versions and downgrades can only work if the configurations are fully compatible, per example,
+current CIDRs are limited to a /112 max for IPv6, if an user configures a /64 to their IPv6 subnets
+in the new model, and IPs are assigned out of the first /112 block, the old allocator based in
+bitmap will not be able to use those IPs creating an inconsistency in the cluster.
+
+It will be required that those Services are recreated to get IP addresses inside the configured
+ranges, for consistency, but there should not be any functional problem in the cluster if the
+Service implementation (kube-proxy, ...) is able to handle those IPs.
+
+Example:
+
+- flags set to 10.0.0.0/20
+- upgrade to 1.25 with alpha gate
+- apiserver create a default ServiceCIDRConfig object for 10.0.0.0/20
+- user creates a new ServiceCIDR for 192.168.1.0/24
+- create a Service which gets 192.168.1.1
+- rollback or disable the gate
+- the apiserver repair loops will generate periodic events informing the user that the Service with the IP allocated
+  is not within the configured range
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: ServiceCIDRConfig
+  - Components depending on the feature gate: kube-apiserver, kube-controller-manager
```

Comment:
What about service implementations (e.g., kube-proxy)? Does it need to have a feature gate?

Reply:
shouldn't - it doesn't need the CIDR

Reply:
no need, kube-proxy doesn't have a notion of service CIDR, it consumes the IPs it sees in the service.spec.ClusterIP ... I think that in this regard is the same as CoreDNS, AFAIK they don't have the service-cidr defined anywhere